pkg/apis/kafka/v1alpha1/reset_offset_lifecycle.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"sync"
+
+	"knative.dev/pkg/apis"
+)
+
+var cs = apis.NewBatchConditionSet(
+	ResetOffsetConditionResetInitiated,
+	ResetOffsetConditionOffsetsCalculated,
+	ResetOffsetConditionConsumerGroupsStopped,
+	ResetOffsetConditionOffsetsUpdated,
+	ResetOffsetConditionConsumerGroupsStarted)
+
+var condSetLock = sync.RWMutex{}
+
+const (
+	// ResetOffsetConditionSucceeded has status True when all sub-conditions below have been set to True.
+	ResetOffsetConditionSucceeded = apis.ConditionSucceeded
+
+	// ResetOffsetConditionResetInitiated has status True when the ResetOffset is being processed
+	// Until that time it is either "Unknown" when first encountered, or "Blocked" if another
+	// ResetOffset is being processed for the same "Ref" resource.
+	ResetOffsetConditionResetInitiated apis.ConditionType = "1-ResetInitiated"
+
+	// ResetOffsetConditionOffsetsCalculated has status True when the individual offsets of each
+	// Partition in the Topic have been calculated from the "offset" indicator in the Spec and
+	// have been verified as valid relative to the persistence window.
+	ResetOffsetConditionOffsetsCalculated apis.ConditionType = "2-OffsetsCalculated"
+
+	// ResetOffsetConditionConsumerGroupsStopped has status True when all of the ConsumerGroups
+	// associated with the referenced object (Subscription, Trigger, etc.) have been stopped.
+	ResetOffsetConditionConsumerGroupsStopped apis.ConditionType = "3-ConsumerGroupsStopped"
```

Comment:
Having the numbers in the condition string doesn't seem very kubernet-ish

Reply:
Yeah I agree it is a bit odd, the reason for that was trying to convey the "order" of the algorithm to the user.  I can remove the numbering but then the status fields are just alphabetical.  For a normal aggregated status case that's not so bad but for this `NewBatchConditionSet` usage it felt wrong.   Still not a big deal - happy to remove the numbers if you still prefer it that way?

Reply:
How kube jobs deal with it?

Reply:
Looks like they have an array of conditions that I suppose could be considered "ordered".  I've removed the numbering and I think we can just start with that.  If it is a problem later (during initial development/testing/usage of ResetOffset) we can look at enhancements (new knative ConditionSet type or something) before final release.